### PR TITLE
[DNNL] activations Inplace support

### PR DESCRIPTION
**Pull request by jczaja**
This PR is extending DNNL inplace mechanism to support activation op. It is follow up to PR#23477.

### CODE REVIEW

**Review by grygielski — APPROVED** [review]
LGTM

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 1]
```diff
@@ -47,8 +52,14 @@ class MKLDNNInplacePassTest {
       op->SetInput("Input", {inputs[0]});
       op->SetInput("Filter", {inputs[1]});
       op->SetInput("Bias", {inputs[2]});
+    } else if (type == "gelu") {
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:55]: What about grouping:
```c++
} else if (std::set<std::string>{"gelu", "leaky_relu", "relu", "tanh"}.count(type)) {
   op->SetInput("X", inputs);
} ...

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 2]
```diff
@@ -90,6 +101,18 @@ class MKLDNNInplacePassTest {
...
+          mkldnn_enabled_op.compare("tanh") == 0);
+    SetOp(&prog, "relu", "relu2", std::vector<std::string>({"l"}),
+          std::vector<std::string>({"m"}),
+          mkldnn_enabled_op.compare("relu") == 0);
+    SetOp(&prog, "leaky_relu", "leaky_relu1", std::vector<std::string>({"m"}),
+          std::vector<std::string>({"n"}),
+          mkldnn_enabled_op.compare("leaky_relu") == 0);
+    SetOp(&prog, "gelu", "gelu1", std::vector<std::string>({"n"}),
+          std::vector<std::string>({"m"}),
+          mkldnn_enabled_op.compare("relu") == 0);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:115]: `relu` -> `gelu`

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 3]
```diff
@@ -90,6 +101,18 @@ class MKLDNNInplacePassTest {
     SetOp(&prog, "relu", "relu2", std::vector<std::string>({"j"}),
           std::vector<std::string>({"k"}),
           mkldnn_enabled_op.compare("softmax") == 0);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:103]: `softmax`->`relu`

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 4]
```diff
@@ -90,6 +101,18 @@ class MKLDNNInplacePassTest {
...
+          std::vector<std::string>({"l"}),
+          mkldnn_enabled_op.compare("tanh") == 0);
+    SetOp(&prog, "relu", "relu2", std::vector<std::string>({"l"}),
+          std::vector<std::string>({"m"}),
+          mkldnn_enabled_op.compare("relu") == 0);
+    SetOp(&prog, "leaky_relu", "leaky_relu1", std::vector<std::string>({"m"}),
+          std::vector<std::string>({"n"}),
+          mkldnn_enabled_op.compare("leaky_relu") == 0);
+    SetOp(&prog, "gelu", "gelu1", std::vector<std::string>({"n"}),
+          std::vector<std::string>({"m"}),
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:114]: `m`->`o`?

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 5]
```diff
@@ -90,6 +101,18 @@ class MKLDNNInplacePassTest {
     SetOp(&prog, "relu", "relu2", std::vector<std::string>({"j"}),
           std::vector<std::string>({"k"}),
           mkldnn_enabled_op.compare("softmax") == 0);
+    SetOp(&prog, "tanh", "tanh1", std::vector<std::string>({"k"}),
+          std::vector<std::string>({"l"}),
+          mkldnn_enabled_op.compare("tanh") == 0);
+    SetOp(&prog, "relu", "relu2", std::vector<std::string>({"l"}),
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:107]: repeated `relu`?

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc [hunk 6]
```diff
@@ -153,6 +172,15 @@ TEST(MKLDNNInplacePass, inplace_elementwise_add) {
...
 }
+TEST(MKLDNNInplacePass, inplace_tanh) {
+  MKLDNNInplacePassTest().MainTest("tanh", false, 1);
+}
+
+TEST(MKLDNNInplacePass, inplace_leaky_relu) {
+  // Input of leaky_relu is used as output of subsequent gelu, so no inplace
+  // cannot be done
+  MKLDNNInplacePassTest().MainTest("leaky_relu", false, 0);
+}
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:183]: What about adding tests for `relu` and `gelu`?

**Review by luotao1 — APPROVED** [review]

**Comment by luotao1** [comment]
Sorry, I miss the comment from @wojtuss, please fix in next PR

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 6 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:183]: Gelu technically is capable of in-place , but PaddlePaddle does not provide InplaceInferer for Gelu (gelu is implemented separatly from other activations) . In other words , currently inplace pass will not work for GELU as inplaceinferer is not present , once in general code there is inplaceinferer for gelu it all starts to work. As for relu/leaky_relu itp. In our code it is the same implementation. So does not matter which activation I choose for UT. Those tests are more for testing certain graphs confguration.

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:55]: goot suggestion , will do that

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:103]: good catch!

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 5 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:107]: Yes, I should have give name relu3. good catch

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:114]: No:) This is intentional. Some aggressive optimization may force paddle to reuse buffers. If one op is having input A that , and that op may run  in-place execution then there may be a situation when we do not do in-place which is when next op is using tensor A as output and this next op cannot do inplace.
So Initial graph is like this:   A -> inplace_to_be_op -> B -> next_op -> A   , then we cannot do:
A -> inplace_to_be_op -> A -> next_op -> A , becouse we would enforce that next_op is working inplace , while it cannot

File: paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/framework/ir/mkldnn/mkldnn_inplace_pass_tester.cc:115]: good catch

**Pull request merged**